base_partner_sequence/models/partner.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2004-2009 Tiny SPRL (<http://tiny.be>).
+# Copyright 2013 initOS GmbH & Co. KG (<http://www.initos.com>).
+# Copyright 2016 Tecnativa - Vicent Cubells
+# Copyright 2016 Camptocamp - Akim Juillerat (<http://www.camptocamp.com>).
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, exceptions, models
+
+
+class ResPartner(models.Model):
+    """Assigns 'ref' from a sequence on creation and copying"""
+
+    _inherit = "res.partner"
+
+    def _get_next_ref(self, vals=None):
+        return self.env["ir.sequence"].next_by_code("res.partner")
+
+    @api.model
+    def create(self, vals):
+        if not vals.get("ref") and self._needsRef(vals=vals):
+            vals["ref"] = self._get_next_ref(vals=vals)
+        return super(ResPartner, self).create(vals)
+
+    def copy(self, default=None):
+        default = default or {}
+        if self._needsRef():
+            default["ref"] = self._get_next_ref()
+        return super(ResPartner, self).copy(default)
+
+    def write(self, vals):
+        for partner in self:
+            partner_vals = vals.copy()
+            if (
+                not partner_vals.get("ref")
+                and partner._needsRef(partner_vals)
+                and not partner.ref
+            ):
+                partner_vals["ref"] = partner._get_next_ref(vals=partner_vals)
+            super(ResPartner, partner).write(partner_vals)
+        return True
+
+    def _needsRef(self, vals=None):
```

Comment:
change _needsRef to _needs_ref. We do not use camelcasing on method names. Change can safely be done, it is an internal method that is not called in any OCA module from outside this module/model.